cpp/src/layout/barnes_hut.hpp

```diff
@@ -89,7 +89,7 @@ void barnes_hut(raft::handle_t const &handle,
   rmm::device_uvector<int> d_childl((nnodes + 1) * 4, stream);
   // FA2 requires degree + 1
   rmm::device_uvector<int> d_massl(nnodes + 1, stream);
-  thrust::fill(rmm::exec_policy(stream)->on(stream), d_massl.begin(), d_massl.end(), 1.f);
+  thrust::fill(rmm::exec_policy(stream), d_massl.begin(), d_massl.end(), 1.f);
```

Comment:
Quick question.  Never noticed this in earlier reviews of this code.

Is it intentional that we use `float` for these values when the graph might have `float` or `double` input types?  Should we be templating these `float` instances to be `weight_t` do avoid the loss of value in conversions, or is it really the case that a graph with `double` weights should be using `float` for these values being computed?

I don't know enough about the context of this algorithm to know the answer, please enlighten.

Reply:
Looks like edge weights are only used in `apply_attraction()` which [discards](https://github.com/rapidsai/cugraph/blob/branch-21.06/cpp/src/layout/fa2_kernels.hpp#L55) the extra precision, so I don't think it matters.

It looks like we should be filling with `1` instead of `1.0f` though, because `d_massl` is a device_uvector of integers.

Reply:
OK.  We should probably have a bigger conversation about this, but I don't think it should hold up progress here.

Could you add the following circa line 94:
```
// FIXME: Review whether these should be "float" or "weight_t"
```

That way we can come back at some point (perhaps when we refactor with the new graph) and revisit this.

Reply:
Two main reasons why I'm against it:

1. Double-precision arithmetic in the force integration kernels probably isn't worth the performance hit.
2. A user could accidentally opt in to a much slower implementation by accidentally using FP64 weights.